Implemented PITR-2

arindamnayak opened the pull request:
Implemented PITR-2
Fixes #6267, fixes #4857 

Here are the changes I did.

- vttablet will accept binlog server parameter
- all those parameters will be used while restoring the tablet and it need a snapshot keyspace.
- from the snapshot keyspace type, we will get the snapshot_time
- we will get the GTID from the snapshot time
- then we will replicate from binlog server upto that GTID
- once we reach that GTID , the slave connection will be closed and reset.

arindamnayak commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,173 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+				"RESET SLAVE ALL",
+			}
+			if err := tm.MysqlDaemon.ExecuteSuperQueryList(ctx, cmds); err != nil {
+				return vterrors.Wrap(err, "failed to reset slave")
```
```suggestion
				return vterrors.Wrap(err, "failed to stop replication")
```

arindamnayak commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,173 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+		return vterrors.Wrap(err, "error while fetching the current gtid position")
+	case <-ctx.Done():
+		log.Warningf("Could not copy till gtid.")
+		return vterrors.Wrap(err, "context timeout while restoring upto specified gtid")
```
```suggestion
		return vterrors.Wrapf(err, "context timeout while restoring upto specified gtid", GTID_VALUE)
```

arindamnayak commented on go/vt/vttablet/tabletserver/schema/engine.go:
```diff
@@ -64,6 +64,11 @@ type Engine struct {
 	notifiers   map[string]notifier
 
+	// SkipMetaCheck skips the metadata about the database and table information
+	SkipMetaCheck bool
```
@sougou , is it ok if we add this flag here?

sougou replied:
Sounds good for now. I see that you're just using it to covert timestamp to gtid. There may be a simpler way to do that without having to use vstream, but we can use this method until we figure that out.

deepthi commented on .github/workflows/docker_test_2.yml:
```diff
@@ -1,10 +1,11 @@
-name: cluster_vtctl_web
+name: docker test 2
```
```suggestion
name: docker_test_2
```

arindamnayak replied:
this is already fixed, marking as resolved.

deepthi commented on go/mysql/flavor_mysql.go:
```diff
@@ -32,7 +32,8 @@ type mysqlFlavor struct{}
 // masterGTIDSet is part of the Flavor interface.
 func (mysqlFlavor) masterGTIDSet(c *Conn) (GTIDSet, error) {
-	qr, err := c.ExecuteFetch("SELECT @@GLOBAL.gtid_executed", 1, false)
+	// making @@global as small case, as the PITR depends on binlog server, which honours only small cased `global` value
```
```suggestion
	// making @@global as lowercase, as PITR depends on binlog server, which honours only lowercase `global` value
```

deepthi commented on go/test/endtoend/recovery/pitr/pitr_test.go:
```diff
@@ -0,0 +1,141 @@
+	_, err := masterTablet.VttabletProcess.QueryTablet(fmt.Sprintf(insertTable, id, productName), keyspaceName, true)
+	require.NoError(t, err)
+	if isSlow {
+		time.Sleep(3 * time.Second)
```
does this have to be 3 seconds? can it be 1 second? As it is the endtoend tests are quite time-consuming.

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -44,6 +51,13 @@ var (
+	binlogPort           = flag.Int("binlog_port", 0, "(PITR restore parameter) port of binlog server.")
+	binlogUser           = flag.String("binlog_user", "", "(PITR restore parameter) username of binlog server.")
+	binlogPwd            = flag.String("binlog_password", "", "(PITR restore parameter) password of binlog server.")
+	timeoutForGTIDLookup = flag.Duration("binlog_lookup_timeout", 60*time.Second, "(PITR restore parameter) timeout for fetching gtid from timestamp.")
```
what do you think of `gtid_lookup_timeout` for this flag name?

arindamnayak replied:
that does not signifies which `gtid` we are referring to, let me know if we should use that.

deepthi replied:
recovery_gtid_lookup_timeout?

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -126,6 +140,14 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
 	if backupManifest != nil {
 		pos = backupManifest.Position
 	}
+	// If restore_to_time is set , then apply the incremental change
```
```suggestion
	// If SnapshotTime is set , then apply the incremental change
```

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
 }
 
+// restoreToTimeFromBinlog restores to the snapshot time of the keyspace
+// currently this works with mysql based database only (as it uses mysql specific queries for restoring
```
```suggestion
// currently this works with mysql based database only (as it uses mysql specific queries for restoring)
```

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+	return nil
+}
+
+// getGTIDFromTimestamp gets the next GTID of the event happened on the timestamp (resore_to_time)
+//
+// it returns the 2 values, 1st one is the after gtid of the timestamp,
+// the 2nd one returns before gtid upto which the replication will be applied
+// 1st can be used directly in the query `START SLAVE UNTIL SQL_BEFORE_GTIDS = ''`
+// 2nd will be used to check if replication completed from the binlog server
```
```suggestion
// getGTIDFromTimestamp computes 2 GTIDs based on restoreTime
// currentPos is the GTID of the first event at or after restoreTime.
// beforePos is the GTID of the last event before restoreTime. This is the GTID upto which replication will be applied
// currentPos can be used directly in the query `START SLAVE UNTIL SQL_BEFORE_GTIDS = ''`
// beforePos will be used to check if replication was able to catch up from the binlog server
```

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+			for _, event := range events {
+				if event.Gtid != "" {
+					currentPos = event.Gtid
+					if event.Timestamp > restoreTime {
```
```suggestion
					if event.Timestamp >= restoreTime {
```

deepthi replied:
I think this is important. In the design we say that we will catch up to the last event before restoreTime.

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+		return val[0], val[1]
+	case <-ctx.Done():
+		log.Warningf("Can't find the GTID from restore time stamp, exiting.")
+		return currentPos, beforeGTIDPos
```
```suggestion
		return currentPos, beforePos
```

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+// the 2nd one returns before gtid upto which the replication will be applied
+// 1st can be used directly in the query `START SLAVE UNTIL SQL_BEFORE_GTIDS = ''`
+// 2nd will be used to check if replication completed from the binlog server
+func (tm *TabletManager) getGTIDFromTimestamp(ctx context.Context, pos mysql.Position, restoreTime int64) (string, string) {
```
Consider naming return values to help with readability. Maybe something like:
```
(currentPos, beforePos string)
```

deepthi commented on go/vt/vttablet/tabletmanager/restore.go:
```diff
@@ -166,6 +188,166 @@ func (tm *TabletManager) restoreDataLocked(ctx context.Context, logger logutil.L
+	}
+}
+
+// catchupToGTID replicates upto specified gtid from binlog server
+//
+// copies the data from binlog server by pointing to as replica
+// waits till all events to gtid replicated
+// once done, it will reset the replication
```
gtid -> GTID

deepthi commented on go/vt/vttablet/tabletmanager/vreplication/replica_connector.go:
```diff
@@ -0,0 +1,103 @@
+	return c.vstreamer.Stream(ctx, startPos, nil, filter, send)
+}
+
+func (c *replicaConnector) VStreamRows(ctx context.Context, query string, lastpk *querypb.QueryResult, send func(*binlogdatapb.VStreamRowsResponse) error) error {
```
Document the interface that this is satisfying.

deepthi commented on go/vt/vttablet/tabletserver/schema/engine.go:
```diff
@@ -308,7 +318,8 @@ func (se *Engine) reload(ctx context.Context) error {
 
 func (se *Engine) mysqlTime(ctx context.Context, conn *connpool.DBConn) (int64, error) {
-	tm, err := conn.Exec(ctx, "select unix_timestamp()", 1, false)
+	// `SELECT UNIX_TIMESTAMP` is in upper case , so that binlog server queries are case sensitive and can respond to this
```
```suggestion
	// `SELECT UNIX_TIMESTAMP` is in uppercase because binlog server queries are case sensitive and expect it to be so.
```

arindamnayak commented:
> Mostly comments and documentation, but there is an important logic issue as well.

I have implemented the suggestion you mentioned.

aquarapid commented on go/mysql/flavor_mysql.go:
```diff
@@ -32,7 +32,8 @@ type mysqlFlavor struct{}
 // masterGTIDSet is part of the Flavor interface.
 func (mysqlFlavor) masterGTIDSet(c *Conn) (GTIDSet, error) {
-	qr, err := c.ExecuteFetch("SELECT @@GLOBAL.gtid_executed", 1, false)
+	// making @@global as lowercase, as the PITR depends on binlog server, which honours only lowercase `global` value
```
Let's just make this:

```
// keep @@global as lowercase, as some servers like the Ripple binlog server only honors a lowercase `global` value
```

Otherwise people might get confused (since this is not PITR-specific code)

arindamnayak replied:
Agreed, I have updated the comment.

deepthi commented:
Can you balance tests between docker_test_1 and docker_test_2?
```
docker test 2 / Docker Test 2 (pull_request) Successful in 19m
docker_test_1 / Docker Test 1 (pull_request) Successful in 9m
```

arindamnayak commented:
> Can you balance tests between docker_test_1 and docker_test_2?
> 
> ```
> docker test 2 / Docker Test 2 (pull_request) Successful in 19m
> docker_test_1 / Docker Test 1 (pull_request) Successful in 9m
> ```

I have balanced the tests

deepthi commented on go/test/endtoend/cluster/cluster_util.go:
```diff
@@ -73,11 +73,11 @@ func GetMasterPosition(t *testing.T, vttablet Vttablet, hostname string) (string
 
-// VerifyRowsInTablet Verify total number of rows in a tablet
-func VerifyRowsInTablet(t *testing.T, vttablet *Vttablet, ksName string, expectedRows int) {
+// VerifyRowsInTabletForTable Verify total number of rows in a tablet
```
```suggestion
// VerifyRowsInTabletForTable Verify total number of rows in a table
```

deepthi commented on go/test/endtoend/recovery/pitr/main_test.go:
```diff
@@ -0,0 +1,135 @@
+		Name: shardName,
+	}
+	for _, tablet := range tablets {
+		//tablet.MysqlctlProcess = cluster.MysqlCtlProcessInstance()
```
remove commented line

deepthi commented on go/test/endtoend/recovery/pitr/main_test.go:
```diff
@@ -0,0 +1,135 @@
+		if proc, err := tablet.MysqlctlProcess.StartProcess(); err != nil {
+			return err
+		} else {
+			// ignore golint warning, we need the else block to use proc
```
don't you need a `//nolint` here? otherwise CI will fail.

arindamnayak replied:
not sure why golangci-lint did not capture this earlier, it is passing without `//nolint`, to be on safe side I have added it.

arindamnayak commented:
> The new logic looks good.
> A few more nits. We can merge once those are addressed.

Implemented the suggested changes and also updated the flag name.

deepthi commented:
Very nice work! This has been a long-standing request.
